weave/trace_server/calls_query_builder.py

```diff
@@ -1281,56 +1297,90 @@ def process_query_to_optimization_sql(
     to avoid filtering out rows where the field is NULL (as they might be part of
     a valid call when combined with other parts).
 
-    Returns an empty string if:
-    1. There are no heavy conditions
-    2. There are OR operations between start-only and end-only fields that cannot be optimized
-
     Performance note: This optimization is critical for queries with heavy fields,
     as it can significantly reduce peak memory by filtering before aggregation.
     """
-    filterable_conditions = [
-        c for c in conditions if c.is_heavy() and not c.is_feedback()
-    ]
-
-    if not filterable_conditions:
+    if not conditions:
         return OptimizationConditions()
 
     def process_operation(
         operation: tsi_query.Operation,
     ) -> Optional[OptimizationConditions]:
         if isinstance(operation, tsi_query.OrOperation):
-            or_conditions = []
+            str_conditions = []
+            datetime_conditions = []
 
             for op in operation.or_:
                 result = process_operand(op)
+                # If any operand can't be optimized, the whole OR can't be optimized
                 if result is None:
-                    return None  # If any operand can't be optimized, the whole OR can't be optimized
-                or_conditions.append(result)
-            if not or_conditions:
-                return None
+                    return None
+                if not result.str_filter_opt_sql and not result.id_datetime_filters_sql:
```

Comment:
i think this is subtly buggy. You could hit a case where strs are not processable, but ids are, in which case you continue down the tree. Then it flips back and forth. The total condition would not be processess-able, but this routine would allow it. The previous short circuiting worked because you only had one applicable condition to check.